ruby_programming/git/advanced_git_lesson.md

```diff
@@ -0,0 +1,182 @@
+
+`rebase -i` is a command which allows you to interactively stop after each commit you're trying to modify, and then make whatever changes you wish. You do have to tell this command which is the last commit you want to edit, for example `git rebase -i HEAD~3` to edit the last three commits. Let's see what this looks like in action. Open up a project you've worked on and type in:
+
+~~~bash
+  $ git log
+  $ git rebase -i HEAD~3
+~~~
+
+You should notice that when rebasing, the commits are listed in opposite order compared to how you see them when you use `log`. Take a minute to look through all of the options the interactive tool offers you. Now let's look at the commit messages at the top of the tool. What I see when I run `git rebase -i HEAD~3` while writing this lesson on Advanced Git Topics is these last three commits:
+
+~~~bash
+pick c85c71dd5 Rewrite Introduction
+pick 4ad2c8df6 Add information on using amend
+pick 729db1da4 Begin work on Remotes
+~~~
+
+If I wanted to edit one of them, I would change the word `pick` to be `edit` for the appropriate commit. If I wanted to remove a commit I would simply remove it from that list, and if I wanted to change the order I would move them around on the list. Let's see what all three would look like! 
+
+~~~bash
+pick 4ad2c8df6 Add information on using amend
+edit c85c71dd5 Rewrite Introduction
+~~~
+
+This would allow me to edit the `Rewrite Introduction` commit, switch its order with the `Add information on using amend` commit, and completely remove the most recent `Begin work on Remotes` commit. When I'm certain I want to do this, I'll save and exit the editor, which will allow me to edit my commit with the following instructions:
+
+~~~bash
+You can amend the commit now, with
+       git commit --amend
+Once youre satisfied with your changes, run
+       git rebase --continue
+~~~
+
+These commit changes would be disasterous for me, so I won't *actually* do it, but it's a great tool to keep in mind for your future projects. That's all there is to it! It seems simple, but this is a very dangerous tool if misused, so be careful. Most importantly, remember to **only rebase commits that exist in your repository, never rebase commits others may work off of.**
+
+
+## Squashing Commits
+
+Using `squash` for your commits is a very handy way of keeping your git history tidy. Squashing makes it easier for others to understand the history of your project. This will become especially important when you begin to merge features in larger projects. What often happens when a feature is merged, is you end up with some visually complex logs of all the changes a feature branch had on a main branch. These commits are important while the feature is in development, but aren't really necessary when looking through the entire history of your main branch.
```

Comment:
Rather than saying this is something that *will* become important in larger projects I think we should clarify that this process *may* be standard on some teams.